archive/2021-07-26-DssSpell.t.sol

```diff
@@ -0,0 +1,2532 @@
+    event Log(string message, address deployer, string contractName);
+
+    // not provided in DSMath
+    function rpow(uint256 x, uint256 n, uint256 b) internal pure returns (uint256 z) {
+      assembly {
+        switch x case 0 {switch n case 0 {z := b} default {z := 0}}
+        default {
+          switch mod(n, 2) case 0 { z := b } default { z := x }
+          let half := div(b, 2)  // for rounding.
+          for { n := div(n, 2) } n { n := div(n,2) } {
+            let xx := mul(x, x)
+            if iszero(eq(div(xx, x), x)) { revert(0,0) }
+            let xxRound := add(xx, half)
+            if lt(xxRound, xx) { revert(0,0) }
+            x := div(xxRound, b)
+            if mod(n,2) {
+              let zx := mul(z, x)
+              if and(iszero(iszero(x)), iszero(eq(div(zx, x), z))) { revert(0,0) }
+              let zxRound := add(zx, half)
+              if lt(zxRound, zx) { revert(0,0) }
+              z := div(zxRound, b)
+            }
+          }
+        }
+      }
+    }
+    function divup(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = add(x, sub(y, 1)) / y;
+    }
+
+    function bytes32ToStr(bytes32 _bytes32) internal pure returns (string memory) {
+        bytes memory bytesArray = new bytes(32);
+        for (uint256 i; i < 32; i++) {
+            bytesArray[i] = _bytes32[i];
+        }
+        return string(bytesArray);
+    }
+
+    // 10^-5 (tenth of a basis point) as a RAY
+    uint256 TOLERANCE = 10 ** 22;
+
+    function yearlyYield(uint256 duty) public pure returns (uint256) {
+        return rpow(duty, (365 * 24 * 60 * 60), RAY);
+    }
+
+    function expectedRate(uint256 percentValue) public pure returns (uint256) {
+        return (10000 + percentValue) * (10 ** 23);
+    }
+
+    function diffCalc(
+        uint256 expectedRate_,
+        uint256 yearlyYield_
+    ) public pure returns (uint256) {
+        return (expectedRate_ > yearlyYield_) ?
+            expectedRate_ - yearlyYield_ : yearlyYield_ - expectedRate_;
+    }
+
+    function castPreviousSpell() internal {
+        SpellLike prevSpell = SpellLike(spellValues.previous_spell);
+        // warp and cast previous spell so values are up-to-date to test against
+        if (prevSpell != SpellLike(0) && !prevSpell.done()) {
+            if (prevSpell.eta() == 0) {
+                vote(address(prevSpell));
+                scheduleWaitAndCast(address(prevSpell));
+            }
+            else {
+                // jump to nextCastTime to be a little more forgiving on the spell execution time
+                hevm.warp(prevSpell.nextCastTime());
+                prevSpell.cast();
+            }
+        }
+    }
+
+    function setUp() public {
+        hevm = Hevm(address(CHEAT_CODE));
+
+        //
+        // Test for spell-specific parameters
+        //
+        spellValues = SpellValues({
+            deployed_spell:                 address(0),        // populate with deployed spell if deployed
```

Comment:
Archived version needs to have the deployed spell address and timestamp.

Reply:
Fixed @kmbarry1